README.md
# One Log

## Install

* Requires installing [stack]() and [maven]() (which implies installing Java)
* To build all the needed components on Unixish systems:

```
$ ./Build.hs build
```

## Run

To run all components on a Unixish system:

```
$ ./Build.hs run
{"log":{"message":"starting java -jar ./pet-store-payment/target/pet-store-payment-1.0-SNAPSHOT.jar server payment-conf.yaml"},"node":"driver"}
{"log":{"message":"starting stack exec pet-store-server -- Dev 9090 localhost 8080"},"node":"driver"}
# ... more logs
^C
```

This spawns the PetStore and Payment processes, wrapping their `stdout` and `stderr` into JSON-formatted log entries.
Typing `Ctrl + C` should stop both processes.

The format of the logs is simple:

* `log`: The output from the process. If this output is already a JSON object, then this field contains an object, otherwise the output text
  is wrapped into an object with field `message`
* `node`: The name of the process from which this log entry comes from. The parent process is named `driver`

## Test

PetStore comes with a *driver* that uses a `Model` to generate tests representing user actions. This driver can be run
against the server:

```
$ stack exec driver-petstore -- localhost 9090
+++ OK, passed 100 tests.
```

## OSQuery

The runner starts 2 system-level services to manage [osquery](https://github.com/facebook/osquery/) system-level logging and querying:

* `osquerys` is a dead-simple Haskell server that will accept any `osqueryd` enrollment request, feed a configuration and dump the logged
items into its `stdout`.
* `osqueryd` is the osquery daemon itself that repeatedly issues requests depending on a configuration file.

The file `osquery.conf` can be modified, it will be read by `osquerys` and used by `osqueryd`.

# Agenda

## Producing Logs

* Have a single logs firehose
* Logs are "machine-readable" first
* Show activity of services through logs

### Base logging of 2 services

* [x] colorised code by service
* [x] base client/server app with textual traces in Haskell
* [x] add another service in Java
* [x] connect the 2 services
* [x] build script
* [x] run script
* [ ] [Y] fix newlines in stack traces from Java

### Structure logging

* [x] structure logging in JSON
* [ ] [A] adding structured logs for payment
* [ ] [Y] tactics for retrofitting "traditional" logs into structured logs

### Aggregate "All" Logs

* From multitail to unilog

* [x] merge logs from 2 services by tailing containers
* [x] add system level logs (osquery)
* [ ] [A] add unix timestamp to aggregated logs
* [x] set node id to be more precise
* [ ] [A] dockerize all containers
* [ ] [Y] refine osquery queries to provide high-level system information
* [ ] [Y] add fswatch
* [ ] [A] add docker logs
* [x] make driver multithreaded
* [ ] ~~petstore model should handle concurrent clients~~

## Consuming Logs

* Using the produced logs for fun and profit

### Cross-Layer Analysis

* [ ] [A] add jaeger server
* [ ] [A/Y] pass tracing ids around services
* [ ] incident analysis: correlate 1 syst evt with several exceptions?
  * kill a service, event notified in docker events, correlate with incident on API
* [ ] flamegraph

### Different Views
* [ ] [Y] Extract Metrics and view in Grafana

### Storage

* [ ] ~~store logs in kafka??~~
* [ ] feed to a file and applky simple jq filtering

### Reconstruct/Patterns

* Biblio
  * [X] Contact François Denis
  * [ ] Reconstructing grammar from samples
* [ ] [A] build a xducer to extract timing information from petstore
  * [ ] use logs to delineate spans inside service
  * [ ] apply rational transducer one logs to produce per-connection analysis
* [ ] reinject logs into application for replay
* [ ] [Y] infer bigrams distribution from logs -> compare
* [ ] [Y] infer sequences from LSTM
  * can we find again the hand-coded xducer from the inferred grammar?

## Future Works

* [ ] infer xducer/language from past logs
* [ ] RNN for infering sequences
* [ ] Is there a way to make Kibana useful?

## Conclusion

* Logs are the semantics of the system

# Plan

## Au commencement

* Une appli client/server
* On lance l'appli et on tail les logs -> du texte de base, des "stack trace"

## Structurer les logs

* Partir de l'appli avec des logs non structurés -> structurer les logs -> avoir un objet log qui a une structure et ne pas dumper du texte arbitraire
* On obtient une matière brute sur laquelle on va pouvoir travailler par la suite -> déjà un résultat utile
* On peut faire des query/filtrage avec jq ! -> richesse de traitement possible, information pas "écrasée"
* On peut parser les logs facilement

## Multiplier les sources de logs (rajouter des traces petit à petit)

### Merger "les" logs des services -> considérer TOUS les logs de ton application

* Ajouter un autre service avec des logs structurés
* Fusionner les 2 logs dans la console en tailant 2 containers
* jaeger/opentracing → traces cross-services

### Merger les logs du système

* Iostat -w 1 → Json
* tools/perf

### Code couleur pour la provenance des logs

### Stocker  les logs

* Les logs c'est pas un coût pour ton application mais une donnée précieuse  → distribuer le stockage
* Pousser les logs dans kafka

## Analyser des logs

### Visualisation

* Reconstruire un état à partir des traces de log → traces → State Machine
  * Pour analyser de la perf structure (flame graph)
  * Pour les tests
  * Reconstruire une "grammaire" à partir des logs, ML, automates probabilistes
* Replay
  * Mais en fait, si on a une appli event sourcé, les logs applicatifs sont gratos et on peut les rejouer


log-controller/ChangeLog.md
# Changelog for one-log

## Unreleased changes


Build.hs
#!/usr/bin/env stack
-- stack runhaskell --package async --package unix --

{-|
A simple script to build and run all components of the ''One Log'' talk
-}

{-# LANGUAGE OverloadedStrings   #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# OPTIONS_GHC -Wall -Werror #-}

import           Color
import           Control.Concurrent.Async
import           Control.Concurrent.Chan
import           Control.Exception        (IOException, catch, finally, throw,
                                           throwIO, try)
import           Control.Monad            (forever, void)
import           Data.Aeson
import qualified Data.ByteString.Char8    as BS8
import qualified Data.ByteString.Lazy     as LBS
import qualified Data.Text.Lazy           as Text
import           Log
import           System.Environment       (getArgs)
import           System.Exit
import           System.IO
import           System.IO.Error          (isEOFError)
import           System.Process

main :: IO ()
main = do
  args <- getArgs
  case args of
    ["run"]   -> doRun
    ["build"] -> doBuild
    _         -> putStrLn "expecting one of 'run' or 'build'"


doBuild :: IO ()
doBuild = do
  runProc "mvn" [ "install" ] "pet-store-payment"
  runProc "stack" [ "test",  "--fast" ]  "."
  runProc "stack" [ "install",  "--fast" ]  "."


doRun :: IO ()
doRun = do
  q  <- newChan
  t  <- tailLogs q stdout

  let procs = [ ("java",  [ "-jar" , "./pet-store-payment/target/pet-store-payment-1.0-SNAPSHOT.jar", "server", "payment-conf.yaml" ],  ".")
              , ( "pet-store-server", [ "Dev" , "9090", "localhost", "8080" ], ".")
              , ( "osquerys",  [ "osquery.conf", "test_server.pem" , "test_server.key", "8088" ],  ".")
              , ( "osqueryd",  [ "--verbose",  "--ephemeral", "--disable_database", "--tls_hostname",  "localhost:8088"
                               , "--tls_server_certs",  "test_server_ca.pem"
                               , "--config_plugin", "tls"
                               , "--config_tls_endpoint",  "/config"
                               , "--logger_tls_endpoint", "/logger"
                               , "--logger_plugin",  "tls"
                               , "--enroll_tls_endpoint", "/enroll", "--enroll_secret_path", "secret.txt"
                               ],  ".")
              ]

  ps <- concat <$> mapM (\ ((name, args, dir), clr) -> spawnProc clr q name args dir) (zip procs genColors)

  void $ waitAnyCancel (t:ps)


runProc :: FilePath -> [ String ] -> FilePath -> IO ()
runProc fp arg dir = do
  putStrLn $ "> " <> fp <> " " <> unwords arg
  res <- createProcess ((proc fp arg) { cwd = Just dir }) >>= \ ( _,_,_, h) -> waitForProcess h
  case res of
    ExitSuccess       -> pure ()
    e@(ExitFailure _) -> exitWith e


spawnProc :: Color -> LogQueue -> FilePath -> [ String ] -> FilePath -> IO [ Async () ]
spawnProc col queue fp arg dir = do
  let name = Text.pack fp

      readOutput :: Handle -> IO (Async ())
      readOutput hd = async $
        forever ((do
                    l <- LBS.fromStrict <$> BS8.hGetLine hd
                    writeChan queue (logEntry name col l))
                 `catch` (\ e  ->
                             if isEOFError e
                             then throw e
                             else writeChan queue (logEntry name col $
                                                   encode $
                                                   object [ "process" .=  name
                                                          , "error" .= show e ])))

  procStart <- try (createProcess ((proc fp arg) { cwd = Just dir, std_in = Inherit, std_out = CreatePipe, std_err = CreatePipe }))

  case procStart of
    Right (_, Just aout, Just aerr, aProc) -> do
      outtid <- readOutput aout
      errtid <- readOutput aerr
      writeChan queue $ logEntry "driver" (Color 200 10 10) (encode $ object [ "message" .= Text.pack ("starting " <> fp <> " " <> unwords arg) ])

      p <- async $ void (waitForProcess aProc) `finally` terminateProcess aProc
      return $ [p, outtid, errtid]
    Left (e :: IOException) -> throwIO $ userError $ "exception launching process " <> fp <> ": " <> show e
    _      -> throwIO $ userError $ "unexpected result from launching process " <> fp
